core/pta/attestation.c

```diff
@@ -0,0 +1,589 @@
+ */
+static struct attestation_data {
+	uint8_t *iv;
+	uint32_t iv_size;
+	uint8_t *pub;
+	uint32_t pub_size;
+	uint8_t *enc_priv;
+	uint32_t enc_priv_size;
+	uint8_t *sig;
+	uint32_t sig_size;
+} att_data;
+
+static TEE_Result free_att_data(TEE_Result reason)
+{
+	att_data.iv_size = 0;
+	att_data.pub_size = 0;
+	att_data.enc_priv_size = 0;
+	att_data.sig_size = 0;
+
+	if (att_data.iv)
+		free(att_data.iv);
+
+	if (att_data.pub)
+		free(att_data.pub);
+
+	if (att_data.enc_priv)
+		free(att_data.enc_priv);
+
+	if (att_data.sig)
+		free(att_data.sig);
+
+	return reason;
+}
+
+static TEE_Result priv_asn1_decode(uint8_t *priv, const size_t priv_len,
+				   uint8_t **d, size_t *d_size)
+{
+	uint32_t size = 0;
+	uint32_t offset = 7; // Beginn at the start of the modulus
+
+	if (priv_len < offset + 3)
+		return TEE_ERROR_SECURITY;
+
+	// Check the header
+	if (priv[0] != 0x30 ||
+	    priv[1] != 0x82 ||
+	    priv[offset] != 0x02 ||
+	    priv[offset + 1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	// Get the size of the first integer aka modulus
+	size = (priv[offset + 2] << 8) | priv[offset + 3];
+	offset += size + 4;
+
+	// Get the size of the second integer aka exponent
+	if (priv[offset] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	size = priv[offset + 1];
+	offset += size + 2;
+
+	if (priv_len < offset + 3)
+		return TEE_ERROR_SECURITY;
+
+	// Decoding d
+	if (priv[offset] != 0x02 || priv[offset + 1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (priv[offset + 2] << 8) | priv[offset + 3];
+
+	if (priv_len < offset + 4 + size)
+		return TEE_ERROR_SECURITY;
+
+	*d = &priv[offset + 4];
+	*d_size = size;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pub_asn1_decode(uint8_t **n, uint8_t **e,
+				  size_t *n_size, size_t *e_size)
+{
+	uint32_t size = 0;
+
+	// Decoding header
+	if (att_data.pub[0] != 0x30 || att_data.pub[1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[2] << 8) | att_data.pub[3];
+	if ((size + 4) != att_data.pub_size)
+		return TEE_ERROR_SECURITY;
+
+	// Decoding first integer aka modulus
+	if (att_data.pub[4] != 0x02 || att_data.pub[5] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[6] << 8) | att_data.pub[7];
+	*n = &att_data.pub[8];
+	*n_size = size;
+
+	// Decoding second integer aka exponent
+	if (att_data.pub[size + 8] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	*e_size = att_data.pub[size + 9];
+	*e = &att_data.pub[size + 10];
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak get_sys_measurement(uint8_t *ptr)
+{
+	// TODO: Dummy function
+	// get_sys_measurement -> do something hardware specific
+	uint8_t sys[] = {0x00, 0x01, 0x02, 0x03,
+			0x04, 0x05, 0x06, 0x07,
+			0x08, 0x09, 0x0a, 0x0b,
+			0x0c, 0x0d, 0x0e, 0x0f,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00 };
+	memcpy(ptr, sys, 32);
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak get_endorsement_key(uint8_t *key)
+{
+	// TODO: Dummy function get_endorsement_key
```

Comment:
Could `huk_subkey_derive()` with a new `enum huk_subkey_usage` be used for this?
It may still make sense to keep this function `__weak`.

Reply:
Yes, it could. But the preferred way would be to keep the endorsement key in hardware, as discussed with ruchi393. In case, there is no appropriate hardware for keeping the key in hardware this is a viable option. But I do not like the idea to include it here, because the stuff with the endorsement key is always hardware dependent and you must think about where to get your endorsement key from.